File: sources/prolong/EMIN/FACT_Blocks/icholRF_apply.cpp
#include "icholRF_apply.h"

void icholRF_apply(const iReg nn, const iExt* __restrict__ iU, const iReg* __restrict__ jU,
                   const rExt* __restrict__ coef_U, const rExt* __restrict__ D_inv,
                   const rExt* __restrict__ vec, rExt* __restrict__ pvec){

   // Initialize pvec
   for (iReg i = 0; i < nn; i++) pvec[i] = 0.0;

   // Forward substitution
   iExt iend = iU[0];
   for (iReg k = 0; k < nn; k++){
      iExt istart = iend;
      iend = iU[k+1];
      pvec[k] = vec[k] - pvec[k];
      for (iExt m = istart; m < iend; m++) pvec[jU[m]] += coef_U[m]*pvec[k];
   }

   // Scale by D_inv
   for (iReg i = 0; i < nn; i++) pvec[i] *= D_inv[i];

   // Backward substitution
   iend = iU[nn];
   for (iReg k = nn-1; k >= 0; k--){
      iExt istart = iend-1;
      iend = iU[k];
      rExt a = 0.0;
      for (iExt m = istart; m >= iend; m--) a += coef_U[m]*pvec[jU[m]];
      pvec[k] -= a;
   }

}

File: sources/Preconditioner/AMG/prolong/EMIN/FACT_Blocks/ICHOL/IcholRF_apply.f90
subroutine IcholRF_apply(nequ,nterm,iU,jU,mat_U,D_inv,vec,pvec)
!-----------------------------------------------------------------------------------------
!
!  Subroutine: IcholRF_apply
!
!  Coded by Carlo Janna
!  February 2011
!
!  Purpose: forward and backward substitution for the application of the root-free
!           Choleski preconditioner ---> pvec:=(U^TDU)^-1*vec
!           (solves the system U^TDU*pvec=vec).
!           Note that D_inv stores the inverse of the diagonal entries
!
!  Variables:
!
!  nequ    : # of equations
!  nterm   : # of non-zero entries in U
!  iU      : topology vector of U
!  jU      : column indices of U
!  mat_U   : coefficients of U
!  D_inv   : inverse of the entries of D
!  vec     : vector to be preconditioned
!  pvec    : preconditioned vector
!
!-----------------------------------------------------------------------------------------

use class_precision

implicit none

! Input variables
integer, intent(in)            :: nequ,nterm
integer, intent(in)            :: iU(nequ+1),jU(nterm)
real(kind=double), intent(in)  :: D_inv(nequ),mat_U(nterm),vec(nequ)

! Output variables
real(kind=double), intent(out) :: pvec(nequ)

! Local variables
integer                        :: k,n1,mm,i,j,m
real(kind=double)              :: a


! Initialize pvec
do i = 1,nequ
   pvec(i) = zero
enddo

! Forward substitution
do k = 1,nequ
   i = iU(k)
   j = iU(k+1) - 1
   pvec(k) = vec(k) - pvec(k)
   do m = i,j
      pvec(jU(m)) = pvec(jU(m)) + mat_U(m)*pvec(k)
   end do
end do

! Scale by D_inv
do k = 1,nequ
   pvec(k) = pvec(k)*D_inv(k)
enddo

! Backward substitution
do k = 1,nequ
   n1 = nequ-k+1
   a = zero
   i = iU(n1)
   j = iU(n1+1) - 1
   do m = i,j
      mm = j-m+i
      a = a + mat_U(mm)*pvec(jU(mm))
   end do
   pvec(n1) = pvec(n1) - a
end do

end subroutine IcholRF_apply
